provisioner/node_pools.go

```diff
@@ -317,6 +337,46 @@ func (p *AWSNodePoolProvisioner) uploadUserDataToS3(userData []byte, bucketName
 	return fmt.Sprintf("s3://%s/%s", bucketName, objectName), nil
 }
 
+func (p *AWSNodePoolProvisioner) renderUploadPKI(templateCtx *templateContext, nodePoolProfilePath string) (string, error) {
+	pkiTemplatePath := path.Join(nodePoolProfilePath, pkiFileName)
+	pkis, err := renderTemplate(templateCtx, pkiTemplatePath)
+	if err != nil {
+		return "", err
+	}
+	pkiFiles := make(map[string]string)
+	if err != nil {
+		return "", err
+	}
+	err = yaml.Unmarshal([]byte(pkis), &pkiFiles)
+	if err != nil {
+		return "", nil
+	}
+	var pkiTar bytes.Buffer
+	gzw := gzip.NewWriter(&pkiTar)
+	tarWriter := tar.NewWriter(gzw)
+	for key, value := range pkiFiles {
+		decoded, err := base64.StdEncoding.DecodeString(value)
+		if err != nil {
+			return "", err
+		}
+		tarWriter.WriteHeader(&tar.Header{
+			Name: key,
+			Size: int64(len(decoded)),
+			Mode: defaultSSLPermissions,
+		})
+		tarWriter.Write(decoded)
+	}
+	err = tarWriter.Close()
+	if err != nil {
+		return "", nil
+	}
+	err = gzw.Close()
```

Comment:
too late, you would not close on error of the for loop for example please use defer

Reply:
same here